### PR TITLE
feat: register issuer oidc provider when registering issuer profile

### LINKED ISSUE #412: issuer adapter: register issuer OIDC client details
Register issuer OIDC client details in `/profile` API.

### PR DESCRIPTION
**Pull request by Moopli**
Fixes #412 

Signed-off-by: Filip Burlacu <filip.burlacu@securekey.com>

### CODE REVIEW

**Comment by Moopli** [comment]
TODO note: have to fix unit tests to add mocking of oidc client setup, and will need some feedback on whether this is going in the right direction.

File: pkg/restapi/issuer/operation/operations.go [hunk 1]
```diff
@@ -259,8 +262,18 @@ type Operation struct {
 	governanceProvider GovernanceProvider
 	routeSvc           routeService
 	messenger          service.Messenger
+	tlsConfig          *tls.Config
+	oidcClientID       string
+	oidcClientSecret   string
+	oidcCallbackURL    string
```
Comment (rolsonquadras) [inline pkg/restapi/issuer/operation/operations.go:268]: Issuer adapter can support multiple issuers. These need to be part of each issuer profile along with issuer oidc url https://github.com/trustbloc/edge-adapter/pull/411/files#diff-d348a1fc808c8f9c0549391446679e7b7f25d533db5c06e982b16be62c98f331R25.

File: pkg/restapi/issuer/operation/operations.go [hunk 2]
```diff
@@ -332,6 +355,13 @@ func (o *Operation) createIssuerProfileHandler(rw http.ResponseWriter, req *http
 		return
 	}
 
+	_, err = o.createOIDCClient(profileData)
+	if err != nil {
+		commhttp.WriteErrorResponseWithLog(rw, http.StatusBadRequest,
```
Comment (llorllale) [inline pkg/restapi/issuer/operation/operations.go:360]: this does not seem to merit a BadRequest code. Looks more like InternalServerError.

File: pkg/restapi/issuer/operation/operations.go [hunk 3]
```diff
@@ -576,6 +606,34 @@ func (o *Operation) createReferenceCredential(token string, profile *issuer.Prof
 	return vcBytes, nil
 }
 
+func (o *Operation) createOIDCClient(profileData *issuer.ProfileData) (oidcClient, error) {
+	if client, present := o.cachedOIDCClients[profileData.ID]; present {
+		return client, nil
+	}
+
+	secret := make([]byte, 256)
```
Comment (llorllale) [inline pkg/restapi/issuer/operation/operations.go:614]: seems like this secret is:

a) not persisted anywhere, and
b) not shared with anyone

**Comment by llorllale** [comment]
I don't understand this flow fully yet.

Can you please map the relevant components to their OAuth/OIDC roles?

File: pkg/restapi/issuer/operation/operations.go (hunk 3 above)
Comment (Moopli) [reply pkg/restapi/issuer/operation/operations.go:614]: I'm still trying to work out how the authorization server should provide the client secret - it needs to be provided as part of registering the issuer-adapter to the issuer.

**Comment by Moopli** [comment]
> I don't understand this flow fully yet.
> 
> Can you please map the relevant components to their OAuth/OIDC roles?

This PR doesn't run through any steps of an Oauth2 flow - it's a pre-step that just provides the issuer adapter with the details it will need when it services an RP request through an RP adapter over didcomm, for which it will make the request to the chosen issuer as usual, but doing so via OIDC and Oauth2 instead. The issuer adapter is the RP in this case, the issuer is the Resource Server, and I assume we'd have ory/hydra running as the issuer's auth server, which the issuer registers in the profile as the OIDC provider URL (json `oidcProvider` in the create profile request).

**Comment by Moopli** [comment]
To address the previous round of comments, I added an encrypted store that the issuer-adapter service will use to store the client secret and ID for each client relationship for each registered OIDC provider. It's an encrypted store to follow best practices for storing client secrets server-side, so it can be persisted even if the issuer-adapter is restarted for any reason, meaning the issuer-adapter doesn't need to re-register with the OIDC provider.

I've added an issuer OIDC provider (using hydra) to the BDD tests, so the issuer adapter demonstrates that it can register as an OIDC client with a production OIDC provider. I've also added a parameter to the create profile request to allow the client ID and secret to be provided by the system administrator directly, in which case the issuer adapter won't attempt dynamic registration.

**Review by rolsonquadras — APPROVED** [review]

**Review by fqutishat — APPROVED** [review]

**Pull request merged**